### PR TITLE
Type CustomLayerInterface renderingMode, onRemove, onAdd, and prerender optional

### PR DESCRIPTION
**Pull request by nreese**
[DefinitelyTyped typed CustomLayerInterface](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/mapbox-gl/index.d.ts#L2240) renderingMode, onRemove, onAdd, and prerender as optional. This PR updates maplibre typings to match DefinitelyTyped.

## Launch Checklist

 - [ ] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [ ] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [ ] Write tests for all new functionality.
 - [ ] Document any changes to public APIs.
 - [ ] Post benchmark scores.
 - [ ] Manually test the debug page.
 - [ ] Suggest a changelog category: bug/feature/docs/etc. or "skip changelog".
 - [ ] Add an entry inside this element for inclusion in the `maplibre-gl-js` changelog: `<changelog></changelog>`.

### CODE REVIEW

File: src/style/style_layer/custom_style_layer.ts [hunk 1]
```diff
@@ -83,7 +83,7 @@ export interface CustomLayerInterface {
     /**
      * @property {string} renderingMode Either `"2d"` or `"3d"`. Defaults to `"2d"`.
      */
-    renderingMode: '2d' | '3d';
+    renderingMode?: '2d' | '3d' | undefined;
```
Comment (drwestco) [inline src/style/style_layer/custom_style_layer.ts:86]: Marking the field as optional via `?` is the same as including `undefined` in the union type.

Comment (nreese) [reply src/style/style_layer/custom_style_layer.ts:86]: From https://kate-dev.medium.com/when-to-use-typescript-optional-property-how-is-it-different-from-declaring-property-as-undefined-2319a0ee1f07#:~:text=Typescript%20%E2%80%9C%3F%E2%80%9D%20operator%20means%20that,its%20value%20may%20be%20undefined.

Screen Shot 2022-03-14 at 1 29 57 PM

Comment (drwestco) [reply src/style/style_layer/custom_style_layer.ts:86]: That's partially correct. `?` implies `| undefined`, though the reverse isn't true:

image

https://www.typescriptlang.org/play?target=1#code/C4TwDgpgBAYg9nKBeKBvAUFLUBOEB2AJhDgJb4DmAsnMQPwBcUA5AEyHNQA+LAzBwG50AXyHoANhGBQAZgibw4QuXAB0eIiXLVaEIA

Comment (drwestco) [reply src/style/style_layer/custom_style_layer.ts:86]: I don't see any cases in the kate-dev blog post where both decorations would be used (all of the examples are one or or the other), and some lint tools explicitly forbid the redundant `| undefined` on optional properties.

https://rules.sonarsource.com/typescript/RSPEC-4782

Comment (nreese) [reply src/style/style_layer/custom_style_layer.ts]: I have changed to `renderingMode?: '2d' | '3d';`

**Comment by HarelM** [comment]
Not sure I understand why onAdd and onRemove should be optional and not mandatory - from API point of view I mean...

**Comment by nreese** [comment]
> Not sure I understand why onAdd and onRemove should be optional and not mandatory - from API point of view I mean...

I am +1 to keep onAdd and onRemove optional. As an API consumer, I found it tedious to provide no-op implementations just to satisfy ever expanding APIs

**Comment by HarelM** [comment]
But that's not an ever expanding example. onAdd and onRemove are kinda the constructor and destructor of the layer, I think that if you choose an empty implementation (`onAdd() {}`) you should do so knowingly.
At least this is how I feel about it.
I've also seen awkward places in the code testing for `if (...onRemove) ...onRemove()` which for me indicates on bad interface design...

**Comment by drwestco** [comment]
They're described in the docs as optional, so marking them as required and especially removing those existence checks would be a breaking change.

**Comment by nreese** [comment]
> But that's not an ever expanding example. onAdd and onRemove are kinda the constructor and destructor of the layer, I think that if you choose an empty implementation (onAdd() {}) you should do so knowingly.
At least this is how I feel about it.
I've also seen awkward places in the code testing for if (...onRemove) ...onRemove() which for me indicates on bad interface design...

Should I create an issue to open up the discussion about changing the existing interfaces? The typescript typings should correctly match the existing API documentation and implementation.

**Comment by nreese** [comment]
created https://github.com/maplibre/maplibre-gl-js/issues/1126.

How about this PR, can this fix move forward to fix existing typescript errors?

**Comment by nreese** [comment]
https://github.com/maplibre/maplibre-gl-js/issues/1126 is not getting a lot of attention. What does that mean for this PR? Can these changes move forward? Updating the typescript types to reflect the current code should be de-coupled from API changes.

**Comment by nreese** [comment]
@HarelM Can this PR move forward? The types should accurately reflect the implementation.

**Comment by HarelM** [comment]
I'll bring it up in the next steering committee meeting to have a wider audience.
Thanks for your patience...

**Review by HarelM — APPROVED** [review]
Steering committee decision: keep backwards compatibility.

**Pull request merged**